ios/ios_testing_references.md

```diff
@@ -68,9 +76,50 @@ class SearchViewModel {
 
 In the above example, the `SearchViewModel` can be initialized with a `NetworkingLayer` to allow a mock `NetworkingLayer` to be injected into the `SearchViewModel`. In this example `NetworkingLayer` is a protocol and as a result the live and mock network layers can both conform to the protocol. The class `DefaultNetworkingLayer` conforms to `NetworkingLayer` and is included as a default initializer. Using protocol is an effective way to implement dependency injection.
 
+### Mocking
+
+As eluded to in the previous section, mocking is a process of creating a dummy or 'mock' object, which will simulate the behavior of a real dependency. This is useful if the real objects are impractical to incorporate into a unit test. Objective-C has access to tools like OCMockito which allow you to automatically generate mock objects at runtime. Unfortunately, because of limitations in the Swift language, you are forced to write your own mocks.
+
+In the above example, we would like to isolate `SearchViewModel` from any real world implementation of `NetworkingLayer`. We do this by implementing a mock version of `NetworkingLayer`, which will run predictably during our testing of `SearchViewModel`:
+
+```swift
+class MockNetworkingLayer: NetworkingLayer {
+    var stubbedResult: Result<[String]> = Result.failure(APIError())
+    func searchFor(searchTerm: String, completionHandler: @escaping (Result<[String]>) -> ()) {
+        DispatchQueue.main.async {
+            completionHandler(self.stubbedResult)
+        }
+    }
+}
+```
+
+We can now test how `SearchViewModel` will behave regardless of how its dependency performs:
+
+```swift
+
+func testSearchWithSuccess() {
+    let mockNetworkingLayer = MockNetworkingLayer()
+    let viewModel = SearchViewModel(networkingLayer: mockNetworkingLayer)
+    mockNetworkingLayer.stubbedResult = Result.success([album1])
+    viewModel.searchFor(searchTerm: "Big")
+    // verify SearchViewModel properly handles successful searches
+    ...
+}
+
+func testSearchWithFailure() {
+    let mockNetworkingLayer = MockNetworkingLayer()
+    let viewModel = SearchViewModel(networkingLayer: mockNetworkingLayer)
+    mockNetworkingLayer.stubbedResult = Result.failure(APIError())
+    viewModel.searchFor(searchTerm: "Big")
+    // verify SearchViewModel properly handles failed searches
+    ...
+}
+
+```
+
```

Comment:
should probably also expand this to include spying which is essentially using a mock but setting it up such that you can verify a particular consumer of a class is calling the methods they are expected to call.

Reply:
Agreed, I will add a section on this.

Reply:
added